### ISSUE TITLE
Osqp back-end

### ISSUE THREAD

**Issue by alaurenzi** [robotology/OpenSoT#108]
Let’s track here the status of the osqp implementation.

I gave a look to the implementation of @EnricoMingo, and I have an observation related to how sparse matrices are handled.

If I understand correctly, at each solver iteration we convert the dense matrices given by the caller into sparse matrices (according to some tolerance). 

I wonder if this is correct for warm starting osqp. I mean, can osqp handle an update of the P and A matrices where not only the values change, but also the sparsity pattern?

**Comment by GiulioRomualdi** [comment]
Hi @alaurenzi 
> I wonder if this is correct for warm starting osqp. I mean, can osqp handle an update of the P and A matrices where not only the values change, but also the sparsity pattern?

Unfortunately at this moment OSQP does not provide any method  to update the hessian and the linear constraint matrix without preserving the sparsity structure of the matrices. This choice is due to the fact that if only the nonzero elements of `P` and `A` change it is possible to avoid a new complete KKT matrix factorization.

So if you want to change the matrices without preserving the sparsity pattern you can throw away the old workspace perform a new `setup`.

For further information please refer to [this](https://github.com/oxfordcontrol/osqp/issues/42).

**Comment by EnricoMingo** [comment]
Hi, I think I will finish a first version of the implementation than we can optimize the code as well as find better options for the solver.

**Comment by EnricoMingo** [comment]
Hi @GiulioRomualdi , thanks for intervening to the issue :+1: 
Maybe it makes sense to initialize all the csc and Eigen::Sparse matrices with  fully dense one (except for the identity part in the constraints and eventually in the cost)? This will preallocate the memory needed, in the worst case, for all these quantities.

**Comment by alaurenzi** [comment]
For our application the hessian matrix can be comsidered dense. The constraint matrix is usually a dense block, stacked on top of a n by n identity matrix that models simple bounds. So, we can use this sparsity pattern (dense upper block plus diagonal block)

**Comment by EnricoMingo** [comment]
So, I think I have solved couple of issues around the OSQPBackEnd, in particular regarding to handling of
bounds and constraints (when bounds are passed but no constraints in particular) and warm/hot start initialization. Now I have a strange error output from the solver which comes out with the test using Cartesian tasks:
```ERROR in quad_form: quad_form matrix is not upper triangular```

**Comment by alaurenzi** [comment]
I believe that you should provide only the non zero elements in the upper part of the hessian!

**Comment by EnricoMingo** [comment]
Yes, I tried actually calling the selfAdjointView<Upper> but it badly segments...

**Comment by alaurenzi** [comment]
Try to assign the sparse matrix with H.triangularView<Eigen::Upper>()

**Comment by EnricoMingo** [comment]
Thanks @alaurenzi , this solved the problem.
